### PR TITLE
[WIP] Spark on Metaflow

### PR DESCRIPTION
**Pull request by hsezhiyan**
- [X] Create a SparkSession utility with a context manager
- [X] Test a sample flow, ensure Python variables can be passed between steps
- [X] Ensure Spark can access S3 objects using an AWS role associated with a service account
- [X] Create a spark decorator that has no utility but replacing the base image of a spark step with an image that contains `spark-submit`

TODO:
- [ ] Move `SparkSessionCreator` outside of the Metaflow codebase
- [ ] Ensure Spark can access S3 objects using either a role or profile

### CODE REVIEW

**Review by talebzeghmi — CHANGES_REQUESTED** [review]
This is not necessary.
We can provide a `get_spark_session()` method that the Flow author imports and uses.
Creating a Spark Session shouldn't be in Metaflow.

example:
```python    
    @step
    def start(self):
        from aip_spark_sdk import get_spark_session
        sc = get_spark_session().sparkContext
```

This requires **NO CHANGE** to Metaflow.

**Comment by hsezhiyan** [comment]
@talebzeghmi `SparkSessionCreator` is akin to `get_spark_session`. The only change would be it to move it to `aip_spark_sdk`. I just updated `SparkSessionCreator` to run the `generate_conf.sh` to produce the Spark config file. `SparkSessionCreator` also allows the user to pass in additional configs.
Removing the Spark decorator would mean no changes to Metaflow at all. I could remove it, but I think a good choice would be to have the decorator specify the spark image to be used. The reason I'm advocating for this is that the `generate_conf.sh` lives in the Spark Python debian images we push to Artifactory, so it's easy to acquire and run them inside  the container. If we choose the conda approach, we'd need to find another way to acquire the `generate_conf.sh` file.

cc @alexlatchford

File: metaflow/tutorials/09-hellokfp/spark_flow.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+class SparkFlow(FlowSpec):
+    """
+    A flow which uses the @spark decorator to run a spark job within a single step.
+    """
+    spark_eventLog_enabled = Parameter(
+        'spark_eventLog_enabled',
+        help='param with default',
+        default="false",
+    )
```
Comment (alexlatchford) [inline metaflow/tutorials/09-hellokfp/spark_flow.py:13]: Can `Parameters` be complex objects (ie. `dict` or `list`) or are they limited to just simple types (ie. `int`, `float`, `bool`, `str`, etc)?

Comment (hsezhiyan) [reply metaflow/tutorials/09-hellokfp/spark_flow.py:13]: They can be `dict` and `list` as well.

**Comment by hsezhiyan** [comment]
Closing this PR:
- SparkSession utility and sample flows will be moved to `aip-spark-sdk`
- Spark decorator will be removed because injecting a new Spark image can mess things up for the AS

**Pull request closed without merging**